### ISSUE TITLE
Treatment of sample weights equal to zero

### ISSUE THREAD

**Issue by charron** [dmlc/xgboost#3787]
My understanding of xgboost and most machine learning models in general is that removing a sample or setting its weight to zero are equivalent as they result in the same loss function.
In the Python example below I tried a few settings which I think should learn the same model by playing with weights and sample removal / duplication.
Most of them are found to be indeed equivalent but the last one (setting weights to zero) learns a different model.

Code:
```python
import numpy as np
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score
from sklearn.datasets import load_breast_cancer

# prepare data
seed = 0
test_size = 0.5
dataset = load_breast_cancer()
X = dataset.data
Y = dataset.target
X_train, X_test, y_train, y_test = train_test_split(
    X, Y, test_size=test_size, random_state=seed)
models = []

# no odd row
X_train_ = X_train[::2]
y_train_ = y_train[::2]
model = XGBClassifier()
model.fit(X_train_, y_train_)
models.append(model)

# twice the even rows with weight 0.5 and no odd row
X_train_ = np.concatenate((X_train[::2], X_train[::2]), axis=0)
y_train_ = np.concatenate((y_train[::2], y_train[::2]), axis=0)
w_train = 0.5 * np.ones_like(y_train_)
model = XGBClassifier()
model.fit(X_train_, y_train_, w_train)
models.append(model)

# twice the even rows and no odd row, scaling the hyperparameters
X_train_ = np.concatenate((X_train[::2], X_train[::2]), axis=0)
y_train_ = np.concatenate((y_train[::2], y_train[::2]), axis=0)
model = XGBClassifier(reg_lambda=2, min_child_weight=2)
model.fit(X_train_, y_train_)
models.append(model)

# weight 2 for even rows and no odd row, scaling the hyperparameters
X_train_ = X_train[::2]
y_train_ = y_train[::2]
w_train = 2 * np.ones_like(y_train_)
model = XGBClassifier(reg_lambda=2, min_child_weight=2)
model.fit(X_train_, y_train_, w_train)
models.append(model)

# zero weight for odd sample rows
w_train = np.ones_like(y_train)
w_train[1::2] = 0
model = XGBClassifier()
model.fit(X_train, y_train, w_train)
models.append(model)

for i, model in enumerate(models):
    print 'AUC {}: {:.4f}'.format(
        i, roc_auc_score(y_test, model.predict_proba(X_test)[:, 1]) * 100.0)
```

Output:
```
AUC 0: 98.8323
AUC 1: 98.8323
AUC 2: 98.8323
AUC 3: 98.8323
AUC 4: 98.8377
```

OS: Ubuntu 16.04
xgboost 0.80

**Comment by hcho3** [comment]
The difference between model 0 and model 3 can be attributed to arbitrary tie-braking: Multiple split candidates have identical gain (change of loss):

Dump of model 0
```
0:[f27<0.147200003] yes=1,no=2,missing=1,gain=94.2990646,cover=35.5
...
```

Dump of model 3
```
0:[f27<0.14774999] yes=1,no=2,missing=1,gain=94.2990646,cover=35.5
...
```

**Comment by charron** [comment]
Makes total sense, thank you very much for the explanation.
